test/dynamo/test_subclasses.py

```diff
@@ -560,18 +552,10 @@ def forward(self):
             exp_graphs=[false_graph, false_graph, true_graph, true_graph],
             exp_frame_count=[1, 1, 2, 2],
             exp_shape_env_guards=[
-                [],
-                # s0 is specialized and guarded in outter shape_env when dynamo checks the guards
-                ["Ne(Piecewise((1, Eq(s0, 5)), (0, True)), 1)"],
-                [
-                    "Ne(Piecewise((1, Eq(s0, 5)), (0, True)), 1)",
-                    "Eq(Piecewise((1, Eq(s0, 3)), (0, True)), 1)",
-                ],
-                [
-                    "Ne(Piecewise((1, Eq(s0, 5)), (0, True)), 1)",
-                    "Eq(Piecewise((1, Eq(s0, 3)), (0, True)), 1)",
-                    "Eq(Piecewise((1, Eq(s0, 3)), (0, True)), 1)",
-                ],
+                ["Ne(s0, 4)"],
+                ["Ne(s0, 4)", "Ne(s0, 5)"],
+                ["Ne(s0, 4)", "Ne(s0, 5)", "Eq(s0, 3)"],
```

Comment:
Ah, the problem is much clearer here.

This looks like you're doing the guard evaluation wrong. What is going on is that you are testing the guards on each cache entry one by one, and each cache test you're doing is polluting the outer environment with a bunch of guards for the other cache entries you tested. This is bad.

What you want is something like, "when testing SymInt in cache entries, first check if the cache entry matches without installing any guard. Only if you actually hit the cache entry, then do you actually apply the guards to the outer environment."

There's not a very good place to put this logic, unfortunately. In particular, I don't want to goop up the pre-existing guard logic with extra handling for SymInt inputs that is unnecessary most of the time. Do you know at compile time if you're doing one of these weird "nested tracing" situations? If so, you might be able to gate guards generation to generate the guards in a different way. I think the easiest way to redo the gen is a wrapper that wraps the check function, and first disables the guard prop to outer env. If it's true, then it reruns the guards, this time propping to the outer env.

I'm stopping the review here.

@masnesral this is actually the same problem you will potentially run into if you don't do the cache lookup tests correctly.

Reply:
> What is going on is that you are testing the guards on each cache entry one by one, and each cache test you're doing is polluting the outer environment with a bunch of guards for the other cache entries you tested.

Yes, you're right.

> I think the easiest way to redo the gen is a wrapper that wraps the check function, and first disables the guard prop to outer env. If it's true, then it reruns the guards, this time propping to the outer env.

That sounds good. I'll try a little bit and get back!